Sources/NIOPerformanceTester/WebSocketFrameEncoderBenchmark.swift

```diff
@@ -0,0 +1,119 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOWebSocket
+
+final class WebSocketFrameEncoderBenchmark {
+    private let channel: EmbeddedChannel
+    private let dataSize: Int
+    private let data: ByteBuffer
+    private let runCount: Int
+    private let dataStrategy: DataStrategy
+    private let cowStrategy: CoWStrategy
+    private var frame: WebSocketFrame?
+
+    init(dataSize: Int, runCount: Int, dataStrategy: DataStrategy, cowStrategy: CoWStrategy) {
+        self.channel = EmbeddedChannel()
+        self.dataSize = dataSize
+        self.runCount = runCount
+        self.dataStrategy = dataStrategy
+        self.cowStrategy = cowStrategy
+        self.data = ByteBufferAllocator().buffer(size: dataSize, dataStrategy: dataStrategy)
+    }
+}
+
+
+extension WebSocketFrameEncoderBenchmark {
+    enum DataStrategy {
+        case spaceAtFront
+        case noSpaceAtFront
+    }
+}
+
+
+extension WebSocketFrameEncoderBenchmark {
+    enum CoWStrategy {
+        case always
+        case never
+    }
+}
+
+
+extension WebSocketFrameEncoderBenchmark: Benchmark {
+    func setUp() throws {
+        // We want the pipeline walk to have some cost.
+        for _ in 0..<3 {
+            try! self.channel.pipeline.addHandler(NoOpOutboundHandler()).wait()
+        }
+        try! self.channel.pipeline.addHandler(WebSocketFrameEncoder()).wait()
+        self.frame = WebSocketFrame(opcode: .binary, data: self.data, extensionData: nil)
+    }
+
+    func tearDown() {
+        _ = try! self.channel.finish()
+    }
+
+    func run() throws -> Int {
+        switch self.cowStrategy {
+        case .always:
+            let frame = self.frame!
+            return self.runWithCoWs(frame: frame)
+        case .never:
+            return self.runWithoutCoWs()
+        }
+    }
+
+    private func runWithCoWs(frame: WebSocketFrame) -> Int {
+        for _ in 0..<self.runCount {
+            self.channel.write(frame, promise: nil)
+        }
+        return 1
```

Comment:
What are we returning here..?

Reply:
For benchmarks in general, it's very important that the optimiser doesn't see through it and decides to not to anything at all. There are many options (for example a `blackhole` function) to stop the optimiser seeing through this being "unnecessary" work. In NIO's perf tests we usually use an `Int` return and the only thing you need to do is to return some integer that depends on the calculation you did.

For example if you received a number of bytes, you could just return `buffer.readableBytes`. In this case, Cory returns `1` here to return that through the `run` function. Arguably, `1` doesn't really depend on the work that's being done here, so we could remove this. The reason this will work is that there's no way the compiler sees through `self.channel.write(...)` doing nothing (it does at least write into the `EmbeddedChannel`'s `outboundBuffer` so it does have an effect.)

Long story short (@Lukasa correct me if I'm wrong): In _this_ case, it's unnecessary but generally returning `Int`s is what we do, usually they should reflect some (it doesn't matter which) property of something you calculate in the work you want to benchmark. In this benchmark, there isn't really anything we can do synchronously so Cory uses `1` :)

Reply:
Nothing really. Just something to avoid the compiler getting too sassy.